Add mandatory pk argument to NonceGen and update test vectors

jonasnick opened the pull request:
Addresses https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2022-October/021000.html

TODO:
- [x] update change log
- [x] mention this in Overview section

sputn1ck commented on bip-musig2.mediawiki:
```diff
@@ -366,7 +367,7 @@ Algorithm ''NonceGen(sk, aggpk, m, extra_in)'':
 * If the optional argument ''extra_in'' is not present:
 ** Let ''extra_in = empty_bytestring''
-* Let ''k<sub>i</sub> = int(hash<sub>MuSig/nonce</sub>(rand || bytes(1, len(aggpk)) || aggpk || m_prefixed || bytes(4, len(extra_in)) || extra_in || bytes(1, i - 1))) mod n'' for ''i = 1,2''
+* Let ''k<sub>i</sub> = int(hash<sub>MuSig/nonce</sub>(rand || bytes(1, len(pk)) || bytes(1, len(aggpk)) || aggpk || m_prefixed || bytes(4, len(extra_in)) || extra_in || bytes(1, i - 1))) mod n'' for ''i = 1,2''
```
Is it correct that only `bytes(1, len(pk))` and not the pk itself is written? python reference code writes the pk as well

jonasnick replied:
Thanks, fixed.

jonasnick commented:
I noticed that we should probably ping the authors of the existing implementations (@llfourn, @NicolasDorier, @roasbeef): This PR is the proposed fix for the vulnerability recently disclosed on the mailing list. It just adds a mandatory `pk` argument to `NonceGen`, which must be the individual public key of the signer (corresponding precisely to the secret key the signer is signing with). This should be easy if, in your implementations, the signer's public key is readily available. If not, there is an alternative fix that does not require adding a new argument but just makes the secret key argument mandatory.

real-or-random commented:
The more I think about this, the more I think we should enforce that the correct secret key is passed by Sign. 

I think that most callers actually don't need the flexibility of setting a tweak after having sent out nonces. But vulnerabilities would still occur if the caller can be "confused" about what tweak to use, so I think the danger is really that the caller screws up. And then, the only way to defend against this in the spec is to check if the pk in Sign matches the pk in NonceGen.

----

Independent observation: At the moment, KeyAggCoeff doesn't even check that the own pk is in the list of all pks... Maybe we should add this check? (In the email to bitcoin-dev we said "2. The signer's public key appears at least two times with different tweaks in the set of public keys that are aggregated." Strictly speaking, this condition is too narrow without the check.)

jonasnick commented:
Made the remaining changes for the mandatory `pk` argument, implemented the idea that appends the `pk` to the secnonce for safety and added the check that the signer's pubkey is in the list of pubkeys. Ready for review now. We should probably also append the pubkey to the secnonce in the test vectors instead of adding them ad-hoc in the test harness.

Roasbeef commented:
In just about all situations by the time we create a session/context, we already know the _signer's_ public key, so this should be pretty easy to tack on. We recently updated our implementation to verify against the new musig2 1.0.0 test vectors, with plans to release a new `lnd` update (v0.15.4) that makes 1.0.0 the default in our experimental RPC server. This is a week or so way, so we can prob add this extra nonce gen arg on as well.

robot-dreams commented on bip-musig2.mediawiki:
```diff
@@ -347,9 +347,10 @@ Algorithm ''ApplyTweak(keygen_ctx, tweak, is_xonly_t)'':
+Algorithm ''NonceGen(sk, pk, aggpk, m, extra_in)'':
 * Inputs:
 ** The secret signing key ''sk'': a 32-byte array (optional argument)
+** The plain public key ''pk'': a 33-byte array
```
Nit: Does it make sense for mandatory argument to be first?

jonasnick replied:
Is there a convention to do this? I went with the convention where usually the secret key is stated before the public key.

robot-dreams commented on bip-musig2.mediawiki:
```diff
@@ -514,6 +515,10 @@ The reference implementation is for demonstration purposes only and not to be us
 
 Implementers must avoid modifying the ''NonceGen'' algorithm without being fully aware of the implications.
+
+One potential pitfall emerges when ''NonceGen'' is altered such that ''k<sub>1</sub>'' and ''k<sub>2</sub>'' are drawn uniformly at random instead of setting them to the output of a ''hash'' function.
```
```suggestion
One potential pitfall emerges when ''NonceGen'' is altered such that ''k<sub>1</sub>'' and ''k<sub>2</sub>'' are drawn uniformly at random instead of setting them to the output of a ''hash'' function that includes ''pk'' as input.
```

robot-dreams commented on bip-musig2/reference.py:
```diff
@@ -537,12 +539,13 @@ def get_value(key):
 
         rand_ = get_value("rand_")
         sk = get_value("sk")
+        pk = get_value("pk") if "pk" in test_case else plain_pk_gen(sk)
```
Nit: Does it make sense to assert `sk` is not `None` in the case where you call `plain_pk_gen`?

jonasnick replied:
I added the pk to the vectors instead (and typed the get_value function).

robot-dreams commented on bip-musig2.mediawiki:
```diff
@@ -78,7 +78,7 @@ The signers start by exchanging public keys and computing an aggregate public ke
 
 '''First broadcast round:'''
-The signers start the signing session by running ''NonceGen'' to compute ''secnonce'' and ''pubnonce''<ref>We treat the 64-byte ''secnonce'' and 66-byte ''pubnonce'' as grammatically singular even though they are serializations of two scalars and two elliptic curve points, respectively. This treatment may be confusing for readers familiar with the MuSig2 paper. However, serialization is a technic ...
+The signers start the signing session by running ''NonceGen'' to compute ''secnonce'' and ''pubnonce''<ref>We treat the ''secnonce'' and ''pubnonce'' as grammatically singular even though they are serializations of two scalars and two elliptic curve points, respectively. This treatment may be confusing for readers familiar with the MuSig2 paper. However, serialization is a technical detail that i ...
```
```suggestion
The signers start the signing session by running ''NonceGen'' to compute ''secnonce'' and ''pubnonce''<ref>We treat the ''secnonce'' and ''pubnonce'' as grammatically singular even though they include serializations of two scalars and two elliptic curve points, respectively. This treatment may be confusing for readers familiar with the MuSig2 paper. However, serialization is a technical detail that is irrelevant for users of MuSig2 interfaces.</ref>.
```

robot-dreams commented on bip-musig2.mediawiki:
```diff
@@ -568,7 +575,10 @@ Algorithm ''DeterministicSign(sk, aggothernonce, pk<sub>1..u</sub>, tweak<sub>1.
 * Let ''R<sub>⁎,1</sub> = k<sub>1</sub>⋅G, R<sub>⁎,2</sub> = k<sub>2</sub>⋅G''
 * Let ''pubnonce = cbytes(R<sub>⁎,2</sub>) || cbytes(R<sub>⁎,2</sub>)''
-* Let ''secnonce = bytes(32, k<sub>1</sub>) || bytes(32, k<sub>2</sub>)''
+* Let ''d = int(sk)''
```
Just confirming, tweaks on `sk` are unrelated to the tweaks provided in the input to `DeterministicSign`, right?

jonasnick replied:
Eorrect, because those tweaks affect the aggregate public key, which goes into nonce coefficient b.

robot-dreams commented on bip-musig2.mediawiki:
```diff
@@ -728,7 +728,8 @@ An exception to this rule is <code>MAJOR</code> version zero (0.y.z) which is fo
 
-* '''0.8.6''' (2022-09-15): Clarify that implementations do not need to support every features and add a test vector for signing with a tweaked key 
+* '''0.9.0''' (2022-10-28): Add mandatory ''pk'' argument to ''NonceGen'', append ''pk'' to ''secnonce'' and check in ''Sign'' that the ''pk'' in ''secnonce'' matches. Update test vectors.
+* '''0.8.6''' (2022-09-15): Clarify that implementations do not need to support every features and add a test vector for signing with a tweaked key
```
```suggestion
* '''0.8.6''' (2022-09-15): Clarify that implementations do not need to support every feature and add a test vector for signing with a tweaked key
```

robot-dreams reviewed (commented):
Looks good on a first pass! Just a few minor suggestions:

jonasnick commented:
Added test vector that would fail in signing when the signer's pk is not in the list (H/T @sputn1ck)

jonasnick commented:
> We should probably also append the pubkey to the secnonce in the test vectors instead of adding them ad-hoc in the test harness.

Done

jonasnick commented:
@Roasbeef 

> In just about all situations by the time we create a session/context, we already know the signer's public key, so this should be pretty easy to tack on.

Thanks, that's good to hear.

real-or-random commented on bip-musig2.mediawiki:
```diff
@@ -713,8 +713,7 @@ An exception to this rule is <code>MAJOR</code> version zero (0.y.z) which is fo
 The <code>MINOR</code> version is incremented whenever the inputs or the output of an algorithm changes in a backward-compatible way or new backward-compatible functionality is added.
 The <code>PATCH</code> version is incremented for other changes that are noteworthy (bug fixes, test vectors, important clarifications, etc.).
 
-* '''1.0.0''' (2022-10-03): Submit draft BIP to the BIPs repository
```
Should we really take back versions?

jonasnick replied:
Strictly speaking this would be require MAJOR version change, since the inputs and outputs are incompatible. We could do view it as MINOR change since this only affects data that remains local to the signer. But OTOH it's also backwards incompatible because it prohibits some applications (in particular applications where you don't know the pubkey at nonce gen time).

One alternative would be that we retroactively rename 1.0.0 to 1.0.0-rc1 and add rc2 with this PR.

jonasnick replied:
> One alternative would be that we retroactively rename 1.0.0 to 1.0.0-rc1 and add rc2 with this PR.

done

real-or-random commented on bip-musig2.mediawiki:
```diff
@@ -132,14 +132,14 @@ Since MuSig2 key aggregation would accept the duplicate keys and not error out,
 In most conceivable scenarios, the assumption that the arguments are different between two executions of ''NonceGen'' is relatively strong, particularly when facing an active attacker.
 
-In some applications, it is beneficial to generate and exchange ''pubnonces'' before the signer's secret key, the final set of signers, or the message to sign is known.
+In some applications, it is beneficial to generate and exchange ''pubnonces'' before the final set of signers, or the message to sign is known.
```
```suggestion
In some applications, it is beneficial to generate and send a ''pubnonce'' before the other signers, their public keys, or the message to sign is known.
```

 - added "public keys" because this makes more sense in a mental model where signers share nonces before they share public keys or use tweaked public keys ("signers" could be fixed but their public keys not yet) 
 - "set of signers" -> "other signers" to avoid confusion because the own public key needs to be fixed
 - "exchange" -> "send" because "other signers" only makes sense when speaking of a specific signer

real-or-random commented on bip-musig2.mediawiki:
```diff
@@ -207,6 +207,11 @@ This specification is written with a focus on clarity.
+The specification also attempts to prevent vulnerabilities that result from accidental mixups of different sessions.
+For example, using a ''pk'' argument in the ''NonceGen'' algorithm that does not match the ''sk'' later provided in ''Sign'' is insecure (see [https://github.com/jonasnick/musig2-tweaking here] for the details).
+Therefore, the specification makes the signer write ''pk'' into the ''secnonce'' array in ''NonceGen'' such that ''pk'' can be checked in ''Sign'' against ''sk''.
+Implementations are free to use other methods to ensure that ''Sign'' receives the output of the right ''NonceGen'' call.
```
I think there's no "right" call, it's just that the keys need to match.
```suggestion
Implementations are free to use other methods to ensure that the secret key passed to ''Sign'' matches the public key given to ''NonceGen''.
```

real-or-random commented on bip-musig2.mediawiki:
```diff
@@ -207,6 +207,11 @@ This specification is written with a focus on clarity.
 In particular, some values are recomputed but can be cached in actual implementations (see [[#signing-flow|Signing Flow]]).
 
+The specification also attempts to prevent vulnerabilities that result from accidental mixups of different sessions.
+For example, using a ''pk'' argument in the ''NonceGen'' algorithm that does not match the ''sk'' later provided in ''Sign'' is insecure (see [https://github.com/jonasnick/musig2-tweaking here] for the details).
```
I was about to suggest something like this but we don't have ''PlainPubKey''?
```suggestion
For example, using a ''pk'' argument in the ''NonceGen'' algorithm that does not match the ''sk'' later provided in ''Sign'' (i.e., ''PlainPubKey(sk) ≠ pk'') is insecure (see [https://github.com/jonasnick/musig2-tweaking here] for the details).
```

Should we add this algorithm, also to make the spec self-contained? Then we could call this in Sign, too. (Strictly speaking it's currently because we don't have a key generation algorithm :D)

jonasnick replied:
We can't really use this in Sign because Sign needs local variable `d' = int(sk)`. We could in principle let PlainPublicKey accept an integer input but that seems suboptimal.

EDIT: actually what I wrote doesn't make much sense, we could just do both `d' = int(sk)` and `pk = PlainPubKey(sk)`.

real-or-random replied:
Let's do this after this PR maybe.

real-or-random commented on bip-musig2.mediawiki:
```diff
@@ -514,6 +524,10 @@ The reference implementation is for demonstration purposes only and not to be us
 
 Implementers must avoid modifying the ''NonceGen'' algorithm without being fully aware of the implications.
+
+One potential pitfall emerges when ''NonceGen'' is altered such that ''k<sub>1</sub>'' and ''k<sub>2</sub>'' are drawn uniformly at random instead of setting them to the output of a ''hash'' function that includes ''pk'' as input.
+The result is that in some situations, an adversary can forge a signature by influencing the individual tweak of a victim signer after seeing the ''pubnonce'' (see [https://github.com/jonasnick/musig2-tweaking here] for the details).
```
Is that even true now that we store the pk in the secnonce? To be honest, I don't see how drawing secnonces randomly is related to the attack at all but I'm probably missing something.

jonasnick replied:
You're right, that's not really clear after the "add public key to secnonce" commit. Should mention that this concern only applies when this is not done (which is what I expect to happen in practice because implementers probably have better data structures for this).

jonasnick replied:
Ok I fixed the paragraph.

real-or-random commented on bip-musig2.mediawiki:
```diff
@@ -514,6 +524,10 @@ The reference implementation is for demonstration purposes only and not to be us
 
 Implementers must avoid modifying the ''NonceGen'' algorithm without being fully aware of the implications.
+
+One potential pitfall emerges when ''NonceGen'' is altered such that ''k<sub>1</sub>'' and ''k<sub>2</sub>'' are drawn uniformly at random instead of setting them to the output of a ''hash'' function that includes ''pk'' as input.
+The result is that in some situations, an adversary can forge a signature by influencing the individual tweak of a victim signer after seeing the ''pubnonce'' (see [https://github.com/jonasnick/musig2-tweaking here] for the details).
```
Nit: I would prefer a footnote to refer to the writeup, like we do for partial signature "forgery."

jonasnick replied:
Hm I'm not a fan of footnotes. The partial signature forgery footnote is different because it explains the attack in detail and it  wouldn't fit the surrounding text.

real-or-random reviewed (commented):
Nice to see that the "real" diff is so small.

I think the fact that we have pk at all in NonceGen also deserves a short explanation (e.g., a short subsection in "Remarks on Security and Correctness"), in particular because we deviate from the paper now. This could be done here in a separate PR if you prefer -- the changes here are good independent of that.

<details>
<summary>Note for attack writeup</summary>
I was about to suggest that I would want to add a note to this explanation (e.g., in a footnote in that explanation) that says something like this: 
> In general, it is sufficient for security that a signer fixes at nonce generation time a deterministic function F(pk_1...u, m)=pk that, given the plain public keys pk1..u of all signers and the message to sign, outputs the public key pk for which this signer will provide the sk when calling Sign with pk_1..u and m. Requiring the signer to pass the public pk already to NonceGen is a simple way to ensure this requirement; it corresponds to the function F(_, _) = pk. 

But now I see that this too far away from this spec, but I'll leave it here. Maybe it's worth a footnote in the writeup of the attack.
</details>

real-or-random commented on bip-musig2.mediawiki:
```diff
@@ -514,6 +519,12 @@ The reference implementation is for demonstration purposes only and not to be us
 
 Implementers must avoid modifying the ''NonceGen'' algorithm without being fully aware of the implications.
+
+One potential pitfall emerges when ''NonceGen'' is altered so the signer's public key ''pk'' is not part of the returned ''secnonce'' array.
+This specification makes the signer write ''pk'' into the ''secnonce'' array (and check ''pk'' against ''sk'' in ''Sign'') to ensure that the public key of the signer is determined before sending out the ''pubnonce''.
+All implementations must guarantee this if they can not rule out that the signer's individual key is tweaked (e.g., derived via BIP32) and that the tweak is known to the adversary (e.g., as in BIP32 unhardened derivation).
+Otherwise, an adversary can forge a signature by influencing the individual tweak of a victim signer after seeing the ''pubnonce'' (see the [https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2022-October/021000.html bitcoin-dev mailing list post] and [https://github.com/jonasnick/musig2-tweaking writeup] for the details).
```
Tried to polish and improve clarity a little bit (at the cost of repeating but I think it's worth it): 
```suggestion
One potential pitfall emerges when ''NonceGen'' is altered so the signer's public key ''pk'' is not part of the returned ''secnonce'' array.
This specification makes the signer provide ''pk'' at nonce generation time and write ''pk'' into the ''secnonce'' array (and check ''pk'' against ''sk'' in ''Sign'') to ensure that the public key of the signer (including all tweaks) is determined before sending out the ''pubnonce''.
If the public key is not determined before sending out the ''pubnonce'', an adversary can forge a signature by influencing the individual tweak of a victim signer after seeing the ''pubnonce'' (see the [https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2022-October/021000.html bitcoin-dev mailing list post] and [https://github.com/jonasnick/musig2-tweaking writeup] for the details).
Thus, implementations must guarantee that the public key of the signer is determined before sending out the ''pubnonce'' if they cannot rule out that the signer's individual key is tweaked (e.g., derived via BIP32) with a tweak known to the adversary (e.g., as in BIP32 unhardened derivation).<ref>The nonce generation algorithm as given in the [https://eprint.iacr.org/2020/1261 MuSig2 paper] does not take ''pk'' as an input (and does not store it in the ''secnonce''). However, the security model considered in the paper does not cover security for signers who tweak their individual key.</ref>
```

Feel free to modify further.

jonasnick replied:
Thanks, pushed (mod "including all tweaks" because the tweaks that happen after key aggregation are not included).

real-or-random reviewed (commented):
ACK mod nit
